Cat food initial

jessLundie opened the pull request:
This branch adds the base functionality for the Cat Food calculator project.

jessLundie committed:
Create cat-food.php

jessLundie committed:
Revert "Create cat-food.php"

This reverts commit 20547dab0ab9ed3432b0045cb3d87cc860c29d47.

jessLundie committed:
Create cat-food.php

jessLundie committed:
folders

Added folder structure

dmsnell commented on cat-food/cat-food.php:
```diff
@@ -0,0 +1,56 @@
+    339,
+    353,
+    366,
+  ); // RER / day by weight 1-20lbs ref. Hill’s Pet Nutrition calculator
```
this is all great! your function does what you want it to. that's important. beyond being _right_ I'm going to leave a few suggestions about being _maintainable_, which is another way of saying "staying right as the definition of right changes over time" (because "right" is a misleading and dynamic term).

to that end the first thing I'd like for you consider in this is how to break apart the functionality into separate pieces. what we have before us is a list of instructions for getting the food given all these very specific input arguments. that amount of detail in the main function can be distracting, noise. for instance, I don't know if it's important to me to understand why these nineteen values exist or why they are what they are in order to understand how to calculate the food intake for my cat.

could you create a function called `base_caloric_intake` that accepts a cat's weight and outputs the appropriate number of kCals required?

```php
function base_caloric_intake($cat_weight) {
	// … stuff, code, etc…
	return …
}
```

In turn `how_much_food` can call into that instead of relying on array-indexing to this very specific list.

```php
$kcalFeed = base_caloric_intake($weightCat) * multiplier;
```

it's hard to leave this point completely isolation without raising another very general theme, which is units. we don't get much opportunity in PHP to specify units for our values so it would help to add them where they are unlisted, such as possibly in `$weightCat` which I think is supposed to represent pounds. we can discuss a novel approach to units at some point in the future after we get through a few rounds of iteration on this code.

jessLundie committed:
Revision cat-food.php

dmsnell commented on cat-food/cat-food.php:
```diff
@@ -0,0 +1,61 @@
+    353,
+    366,
+  );
+	return $kcalBase[$cat_weight-1];
```
nice. now we don't have to worry about the specifics of array indexing in the main function.

dmsnell replied:
nitpick: might want to circle back on the formatting/indentation. it doesn't change the code functionality (in PHP) but it can be confusing to read

jessLundie replied:
I think I fixed the formatting/indentation issues, but please let me know if I missed something! I'm not sure if I understand the correct paradigm, but I did some looking online and tried to compare to correct examples.

dmsnell commented on cat-food/cat-food.php:
```diff
@@ -0,0 +1,61 @@
+  $qtyDry = round ((($percentDry * $kcalFeed) / $kcalDry) / $tbsConv, 2); // calculate qty dry food(tbs)
+  $qtyCan = round (($percentWet * $kcalFeed) / $kcalWet, 2); // calculate qty wet food (cans)
+
+  if (($cat_weight >= 1) && ($cat_weight <= 20)) {
```
this is awesome checking the arguments, but notice how we're checking the input bounds only _after_ running all of our computation. this would generally be categorized as a "precondition" for the function, and we should get in the habit of _validating_ our inputs at the start before letting those invalid values run through the system, so to speak.

in some programming languages we can actually rely on the compiler to guarantee this for us, but in PHP we have to do it ourselves. see if you can move this up to the top of the function. notice also now that this constraint isn't so much given by this function anymore, but rather by `base_caloric_intake`. it may be possible to remove the hard-encoding of the limits in this function and put them in that one, where the constraint truly comes from. what do you think should happen if you try to call `base_caloric_intake` with a 40lb cat?

jessLundie replied:
Ah, of course! Thank you :) 

I made some changes that appear to work well, but I'm not sure if it's advisable / correct to call `base_caloric_intake` within `how_much_food` and then do a callback to `how_much_food` within base caloric intake. I ran into trouble testing this because our current testing method calls `how_much_food` directly. 

But I do think that what should happen when I call `base_caloric_intake` is:

1. Validate conditions
2. Only run `how_much_food` if `$cat_weight` is between 1-20. 

Is there a better way to do this without putting `how_much_food` entirely within the `$cat_weight` if/else?

dmsnell replied:
> Is there a better way to do this without putting how_much_food entirely within the $cat_weight if/else?

yes of course! there are 13,753 different ways 😉

> I'm not sure if it's advisable / correct to call base_caloric_intake within how_much_food and then do a callback to how_much_food within base caloric intake

this is good intuition, and you should listen to it.

---

try and step back and think again _only_ from the perspective of `base_caloric_intake` - that function has a clear API: it takes a count of pounds as input and returns a count of calories. So what if we don't give it a valid pound range as input?

there's a special value in PHP and in most languages for this kind of "there is no valid return from this function" situation. can you think of what that could be? (technically there are a couple ways to do this)

dmsnell replied:
lovely update! `null`'s the special thing. right now though look at the visual shape of the function

```php
if (…) {














} else {
    
}
```

what if we stored the return value from `base_caloric_intake` into a variable as before, but then only checked if it's invalid, and then doing an "early abort" in that case, leaving it the same. the new visual picture could become

```php
$… = …
if (…) {
    return …
}








```

this is a really _really_ common pattern in the functions we call. "early abort" is also one of the styles that Automattic tends to prefer in our projects. it's nice because once we're past the error-checking we don't have to think about it anymore. with the first visual structure we have to hold on to that thing in our head as long as we are indented, and keep everything else going on inside the indent as well.

jessLundie replied:
Ah, great! I had considered that as an option but wasn't sure on pros and cons of that method. I hadn't realized I could structure that in a way that would omit the `else` entirely. That's way better! Update incoming.

jessLundie committed:
Update cat-food.php

Indent revisions & relocation of condition check

jessLundie committed:
Update cat-food.php

Return null for failed condition check in base_caloric_intake, and condition check for null on how_much_food

jessLundie committed:
Update cat-food.php

Validate weight check via variable.

dmsnell commented on cat-food/cat-food.php:
```diff
@@ -0,0 +1,67 @@
+  if (($cat_weight >= 1) && ($cat_weight <= 20)) {
+    return $kcalBase[$cat_weight-1];
+  } else {
+    echo "Please enter a weight between 1 and 20.";
```
we don't need to change this here, but let's talk about it later

dmsnell commented on cat-food/cat-food.php:
```diff
@@ -0,0 +1,67 @@
+    366,
+  );
+
+  if (($cat_weight >= 1) && ($cat_weight <= 20)) {
```
note that we can apply the same kind of _pre-condition_ early-abort inside this function too! it's turtles all the way down 😄

jessLundie replied:
Question on this - how can we do an early abort on this if we still want it to echo the validation message on the `else`? Or in this case, would we move that message to somewhere else?

dmsnell replied:
if I understand you properly then the answer might be simpler than you think. early-aborts tend to check for rule _breaks_, which are directly the logical inverse of rule _matches_.

imagine you have this rule:
```php
$is_allowed_weight = $cat_weight >= 1 && $cat_weight <= 20;
```

all we've done is captured the exact code you wrote in a variable in order to talk about it with a given name. the current code reads like this

```php
if ($is_allowed_weight) {
    return …
} else {
    echo "WRONG";
    return null;
} 
```

we can invert the logical condition and that flips which of the if/else clauses runs. it truly flips them, so this is the same thing written with the inversion/negation

```php
if (!$is_allowed_weight) {
    echo "WRONG";
    return null;
} else {
    return …
}
```

at this point we can see that if we `return` in the first if clause then we'll never reach the rest of the function in that case. this means that the stuff in the `else` represents everything that happens if our preconditions _aren't wrong_. we can eliminate the `else`

```php
if (!$is_allowed_weight) {
    echo "WRONG";
    return null;
}

return …
```

we've removed a level of indentation and the "early abort" matches the mental model: "if the conditions aren't met then return `null` and skip the rest of the function. otherwise carry along as normal."

if we weren't enthused about creating the extra name/variable for the precondition we can do the negation inline. we can simplify it with some principles called DeMorgan's laws for inverting logical clauses

```php
if (!$is_allowed_weight)                          // start with our name
if (!($is_allowed_weight))                        // this is what it's really doing
if (!(($cat_weight >= 1) && ($cat_weight <= 20))) // substitute what we put in the variable
if (($cat_weight < 1) || ($cat_weight > 20))      // flip via DeMorgan's laws by swapping `&&`s and `||`s, flipping comparisons
if ($cat_weight < 1 || $cat_weight > 20) {        // simplify by removing unnecessary parenthetical groups
    return null;
}
```

dmsnell commented on cat-food/cat-food.php:
```diff
@@ -0,0 +1,67 @@
+      "Weight gain for cat" => 1.8,
+      "Kitten 0 to 4 months" => 2.5,
+      "Kitten 4 months to a year" => 2,
+    ); // RER multiplier based on specific factors
```
so the next big thing here is to apply what we just did with `base_caloric_intake`. turns out that there are many occasions when creating a function can help us give a name to a thing and then take incidental details (how exactly the goal maps to a multiplier) out into their own place.

there will be discussion and argument about this at times. it's a subjective call because in some ways this makes it harder to understand what exactly is happening in the code. you have to jump into another function to understand the mapping from weight goals to multipliers. the benefits though are valuable as well:

 - we can more easily test the behavior of the multiplier selection _in isolation_, in what we call "unit tests" since we don't have to test the entire surrounding functionality in order to test the detail.
 - a clear way to adjust the details. do we want to have to think about calculating food in the pure case where we only want to adjust diet multipliers?
 - separated functionality gives us options (as we will discuss later) for getting the output with different inputs. right now we're passing in exact strings. this may or may not be the way we end up _exposing this interface_ and being able to change one function does make it easier altogether to change than when we have to change the braids inside an existing rope, so to speak.

jessLundie committed:
Update cat-food.php

Revised for WordPress standards * separated out diet_type function.

jessLundie committed:
Update cat-food.php

Fix extra else and resulting syntax errors.

jessLundie committed:
Naming changes, add functions

Naming changes, early abort fixes, and additional front-end functions added for eventual display.

jessLundie committed:
Update cat-food-calculator.php

Add shortcode / form code

dmsnell reviewed (approved):
Let's merge this working bit of code and iterate in future PRs!
